core/predicate_check.go

```diff
@@ -18,50 +18,90 @@ import (
+}
+
+// CheckTxPredicates verifies the predicates of a transaction and returns the
+// results.
+//
+// Returning an error invalidates the transaction.
+func CheckTxPredicates(
+	rules params.Rules,
+	predicateContext *precompileconfig.PredicateContext,
+	tx *types.Transaction,
+) (predicate.PrecompileResults, error) {
+	// Check that the transaction can cover its IntrinsicGas, including the gas
+	// required by the predicate, before verifying the predicate.
+	accessList := tx.AccessList()
+	intrinsicGas, err := IntrinsicGas(tx.Data(), accessList, tx.To() == nil, rules)
 	if err != nil {
 		return nil, err
 	}
-	if tx.Gas() < intrinsicGas {
-		return nil, fmt.Errorf("%w for predicate verification (%d) < intrinsic gas (%d)", ErrIntrinsicGas, tx.Gas(), intrinsicGas)
+	if gas := tx.Gas(); gas < intrinsicGas {
+		return nil, fmt.Errorf("%w for predicate verification (%d) < intrinsic gas (%d)",
+			ErrIntrinsicGas,
+			gas,
+			intrinsicGas,
+		)
 	}
 
 	rulesExtra := params.GetRulesExtra(rules)
-	predicateResults := make(predicate.PrecompileResults)
 	// Short circuit early if there are no precompile predicates to verify
 	if !rulesExtra.PredicatersExist() {
-		return predicateResults, nil
+		return nil, nil
 	}
 
 	// Prepare the predicate storage slots from the transaction's access list
-	predicateArguments := predicate.FromAccessList(rulesExtra, tx.AccessList())
+	predicateArguments := predicate.FromAccessList(rulesExtra, accessList)
 
-	// If there are no predicates to verify, return early and skip requiring the proposervm block
-	// context to be populated.
+	// If there are no predicates to verify, return early and skip requiring the
+	// proposervm block context to be populated.
 	if len(predicateArguments) == 0 {
-		return predicateResults, nil
+		return nil, nil
 	}
 
 	if predicateContext == nil || predicateContext.ProposerVMBlockCtx == nil {
 		return nil, ErrMissingPredicateContext
 	}
 
+	var (
+		txHash           = tx.Hash()
+		predicateResults = make(predicate.PrecompileResults, len(predicateArguments))
+	)
 	for address, predicates := range predicateArguments {
-		// Since [address] is only added to [predicateArguments] when there's a valid predicate in the ruleset
-		// there's no need to check if the predicate exists here.
-		rules := params.GetRulesExtra(rules)
-		predicaterContract := rules.Predicaters[address]
+		// Since address is only added to predicateArguments when there's a
+		// valid predicate in the ruleset there's no need to check if the
+		// predicate exists here.
+		predicaterContract := rulesExtra.Predicaters[address]
 		bitset := set.NewBits()
 		for i, predicate := range predicates {
 			if err := predicaterContract.VerifyPredicate(predicateContext, predicate); err != nil {
 				bitset.Add(i)
 			}
 		}
-		log.Debug("predicate verify", "tx", tx.Hash(), "address", address, "res", bitset.Bytes())
+		log.Debug("predicate verify",
+			"tx", txHash,
+			"address", address,
+			"res", bitset.String(),
```

Comment:
I think `String()` makes more sense than `Bytes()` since this is going into the logger.